src/butil/iobuf.cpp

```diff
@@ -1211,6 +1211,9 @@ int IOBuf::append_user_data(void* data, size_t size, void (*deleter)(void*)) {
         LOG(FATAL) << "data_size=" << size << " is too large";
         return -1;
     }
+    if (!size) {
+        return 0;
```

Comment:
Maybe we need to call deleter here?

Reply:
I still think size==0 should return -1 rather than 0. If it is failed, it means IOBuf does not take this Block at all. It is the caller's responsibility to delete the buffer.

Reply:
@Tuvie Makes sense. Does return -1 break backward compatibility? 
Also return 0 && call deleter means that IOBuf takes the responsibility to free the buff. This is consistent with the behavior before this PR.

Reply:
@wwbmmm Could you have a look and decide which is better?

Reply:
I think call deleter and return 0 is ok.

Reply:
Thanks. Maybe we can merge now 😊

Reply:
Shouldn't we move the deleter null check and assignment below before this change? Otherwise the data may be leaked if user doesn't pass deleter.

Reply:
@zyearn Yes, before this change,  when size equals 0, `append_user_data` relies on the default `free` to release user memory if deleter not specified. We should maintain the semantics here. CC @wwbmmm

Reply:
LGTM